mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DefaultDaoReturnTypeParser.java

```diff
@@ -0,0 +1,229 @@
+        Element returnElement = declaredReturnType.asElement();
+        for (Map.Entry<Class<?>, DaoReturnTypeKind> entry : entityContainerMatches.entrySet()) {
+          Class<?> containerClass = entry.getKey();
+          if (context.getClassUtils().isSame(returnElement, containerClass)) {
+            return new DaoReturnType(entry.getValue(), entityElement);
+          }
+        }
+      }
+
+      if (context.getClassUtils().isFuture(declaredReturnType)) {
+        TypeMirror typeArgumentMirror = declaredReturnType.getTypeArguments().get(0);
+
+        // future of a simple class? e.g. CompletableFuture<Boolean>
+        for (Map.Entry<Class<?>, DaoReturnType> entry : futureOfClassMatches.entrySet()) {
+          Class<?> simpleClassArgument = entry.getKey();
+          if (context.getClassUtils().isSame(typeArgumentMirror, simpleClassArgument)) {
+            return entry.getValue();
+          }
+        }
+
+        // Note that futures of entities (e.g. CompletionStage<Product>) are already covered by the
+        // "entity container" check above
+
+        // future of entity container? e.g. CompletionStage<Optional<Product>>
+        if (typeArgumentMirror.getKind() == TypeKind.DECLARED) {
+          DeclaredType declaredTypeArgument = (DeclaredType) typeArgumentMirror;
+          if (declaredTypeArgument.getTypeArguments().size() == 1
+              && (entityElement =
+                      EntityUtils.typeArgumentAsEntityElement(typeArgumentMirror, typeParameters))
+                  != null) {
+            Element typeArgumentElement = declaredTypeArgument.asElement();
+            for (Map.Entry<Class<?>, DaoReturnTypeKind> entry :
+                futureOfEntityContainerMatches.entrySet()) {
+              Class<?> containerClass = entry.getKey();
+              if (context.getClassUtils().isSame(typeArgumentElement, containerClass)) {
+                return new DaoReturnType(entry.getValue(), entityElement);
+              }
+            }
+          }
+        }
+      }
+    }
+
+    if (returnTypeMirror.getKind() == TypeKind.TYPEVAR) {
+
+      // entity class? e.g. Product
+      TypeElement entityElement;
+      if ((entityElement = EntityUtils.asEntityElement(returnTypeMirror, typeParameters)) != null) {
+        return new DaoReturnType(DefaultDaoReturnTypeKind.ENTITY, entityElement);
+      }
+
+      // simple class? e.g. Boolean
+      TypeVariable typeVariable = ((TypeVariable) returnTypeMirror);
+      Name name = typeVariable.asElement().getSimpleName();
+      TypeElement element = typeParameters.get(name);
+      if (element != null) {
+        for (Map.Entry<Class<?>, DaoReturnType> entry : classMatches.entrySet()) {
+          Class<?> simpleClass = entry.getKey();
+          if (context.getClassUtils().isSame(element, simpleClass)) {
+            return entry.getValue();
+          }
+        }
+      }
+
+      // DAO parameterization by more complex types (futures, containers...) is not supported
```

Comment:
I think we've discussed that already in #1250. I don't think this will be a common need, but we can always revisit later if needed.

Reply:
:+1: although just to make sure I understand the comment correctly, these are for parameterization of types that we don't already handle, such as `CompletableFuture`, `Optional`, etc, right?  I think this is fine, especially with this PR adding the capability to add your own return type parsing logic.

Reply:
This wouldn't work:
```java
interface BaseDao<SelectResultT> {
  @Select
  SelectResultT findAll();
}
@Dao
interface ProductDao extends BaseDao<PagingIterable<Product>> {
}
```
Because the `typeParameters` map has `TypeElement` values. We'd need to store type mirrors in order to resolve `SelectResultT` correctly.
But that's a silly example, I don't think it will come up in practice.

Reply:
Ahh I see, not a scenario I considered and seems unlikely :+1: